signer/threshold_signer_soft.go

```diff
@@ -0,0 +1,354 @@
+		return HrsMetadata{}, err
+	}
+
+	meta = HrsMetadata{
+		Secret: secret,
+		Peers:  make([]PeerMetadata, softSigner.Total),
+	}
+
+	// split this secret with shamirs
+	// !! dealt shares need to be saved because dealing produces different shares each time!
+	meta.DealtShares = tsed25519.DealShares(meta.Secret, softSigner.Threshold, softSigner.Total)
+
+	softSigner.HrsMeta[hrsKey] = meta
+
+	return meta, nil
+}
+
+// Get the ephemeral secret part for an ephemeral share
+// The ephemeral secret part is encrypted for the receiver
+// Implements ThresholdSigner
+func (softSigner *ThresholdSignerSoft) GetEphemeralSecretPart(
+	req CosignerGetEphemeralSecretPartRequest, m *LastSignStateStruct, peers map[int]CosignerPeer) (
+	CosignerEphemeralSecretPart, error) {
+
+	res := CosignerEphemeralSecretPart{}
+
+	// protects the meta map
+	m.LastSignStateMutex.Lock()
+	defer m.LastSignStateMutex.Unlock()
+
+	hrst := HRSTKey{
+		Height:    req.Height,
+		Round:     req.Round,
+		Step:      req.Step,
+		Timestamp: req.Timestamp.UnixNano(),
+	}
+
+	meta, ok := softSigner.HrsMeta[hrst]
+	// generate metadata placeholder
+	if !ok {
+		newMeta, err := softSigner.DealShares(CosignerGetEphemeralSecretPartRequest{
+			Height:    req.Height,
+			Round:     req.Round,
+			Step:      req.Step,
+			Timestamp: req.Timestamp,
+		})
+
+		if err != nil {
+			return res, err
+		}
+
+		meta = newMeta
+		softSigner.HrsMeta[hrst] = meta
+	}
+
+	ourEphPublicKey := tsed25519.ScalarMultiplyBase(meta.Secret)
+
+	// set our values
+	meta.Peers[softSigner.Key.ID-1].Share = meta.DealtShares[softSigner.Key.ID-1]
+	meta.Peers[softSigner.Key.ID-1].EphemeralSecretPublicKey = ourEphPublicKey
+
+	// grab the peer info for the ID being requested
+	peer, ok := peers[req.ID]
+	if !ok {
+		return res, errors.New("unknown peer ID")
+	}
+
+	sharePart := meta.DealtShares[req.ID-1]
+
+	// use RSA public to encrypt user's share part
+	encrypted, err := rsa.EncryptOAEP(sha256.New(), rand.Reader, &peer.PublicKey, sharePart, nil)
+	if err != nil {
+		return res, err
+	}
+
+	res.SourceID = softSigner.Key.ID
+	res.SourceEphemeralSecretPublicKey = ourEphPublicKey
+	res.EncryptedSharePart = encrypted
+
+	// sign the response payload with our private key
+	// cosigners can verify the signature to confirm sender validity
+	{
+		jsonBytes, err := tmJson.Marshal(res)
+
+		if err != nil {
+			return res, err
+		}
+
+		digest := sha256.Sum256(jsonBytes)
+		signature, err := rsa.SignPSS(rand.Reader, &softSigner.Key.RSAKey, crypto.SHA256, digest[:], nil)
+		if err != nil {
+			return res, err
+		}
+
+		res.SourceSig = signature
+	}
+
+	res.DestinationID = req.ID
+
+	return res, nil
+}
+
+// Store an ephemeral secret share part provided by another cosigner (signer)
+// Implements ThresholdSigner
+func (softSigner *ThresholdSignerSoft) SetEphemeralSecretPart(
+	req CosignerSetEphemeralSecretPartRequest, m *LastSignStateStruct, peers map[int]CosignerPeer) error {
+
+	// Verify the source signature
+	{
```

Comment:
Scope-introducing blocks like this are very unusual in Go.

Reply:
It looks strange TBH. Never seen it before either. But this is just a refactor of the current implementation as well. 

https://github.com/strangelove-ventures/horcrux/blob/a3470b5f1a00bbd8af6b0f63ef5f4f119a2cc8aa/signer/local_cosigner.go#L396